### DISCUSSION TITLE
Intermittent lag spikes on draw thread

### DISCUSSION THREAD

**Discussion by joshuk** [ppy/osu#26090] [Q&A] [answered]
### Type

Performance

### Bug description

Before we start, I'm not 100% sure that the title is fully correct, however it's the most accurate representation of the issue I'm having.

After booting up the game today (both before and after applying the most recent update) my game seems severely limited in terms of FPS (the counter seems to say 200-250, however it feels closer to 30-60). However, I am unsure whether the FPS is consistently lower than what is shown in the FPS counter, or whether the game is dropping frames instead (on the video you can see the FPS counter seemingly spike downwards on occasion).

I have attached a video of the performance while playing the map with both a screen recording and an external camera recording, both recording at 60FPS. The FPS issues are most noticable on spinners, where you'll see the cursor seemingly jump from one place to another on both cams.

I am using a 144hz monitor with the '2x refresh rate' option selected in the osu!lazer settings. My PC _should be_ fully capabale of running osu!lazer, I can post the specs if needed.

It should be noted that this issue seems to be intermittent for me when it comes to playing lazer. Some days the game seems to be completely smooth, other times I face issues like this. In comparison, stable is almost entirely smooth all the time. I switched back to stable after recording this video which performed perfectly (I also recorded this session if you would like a comparison).

Happy to provide any more information if needed, I'm sorry about the somewhat vague issue.

### Screenshots or videos

https://github.com/ppy/osu/assets/25204957/0b2f4068-6c25-4761-be94-df7333421d32

### Version

osu! 2023.1224.0-lazer

### Logs

[compressed-logs.zip](https://github.com/ppy/osu/files/13760258/compressed-logs.zip)

**Comment by bdach** [comment]
Please upload a video of this with frame graph showing (ctrl-f11 twice).

**Reply by joshuk** [reply]
Apologies, didn't know that this existed when making the issue. Here you are:

https://github.com/ppy/osu/assets/25204957/e4a28419-5113-4493-b2b8-8bad30fc3e52

**Reply by bdach** [reply]
So... whatever the draw thread is doing there is the cause of this, presumably. But I've never seen this sort of thing before.

Can you start with checking if there are any driver updates for your graphics card?

**Reply by joshuk** [reply]
Right, I have installed the latest Nvidia driver and it appears that it has fixed the issue!

I'll mark this as an answer for now, but as I said previously this issue seems to be intermittent so I'll come back if it happens again. Thanks!

**Reply by peppy** [reply]
Phew!

**Reply by joshuk** [reply]
Hey so remember when I said I'd come back if it happened again 🙃 This one is actually a two parter.

Initially I ran into this issue while playing, but much less consistently than before. It's barely visible on this video below, but is somewhat distracting while playing:

https://github.com/ppy/osu/assets/25204957/290ccb98-5e94-4981-bf1f-736538f3792f

One other thing I've noticed with this is that it takes a very long time to tab in properly after tabbing out for a few minutes (e.g to trim the video above). This seems to only be on the draw thread though, since if I move my cursor around I can still hear sounds from hovering over buttons etc. It also seems to be impossible to close the game normally when it's in this state (even if I close it using keyboard shortcuts, which closes the window itself, the process is still running in Task Manager). Either that or it takes a very long time to actually fully close.

After ending the task in Task Manager and restarting the game to get the log files it appears that it is now doing it much more often than before:

https://github.com/ppy/osu/assets/25204957/c9f020d8-5944-4f5c-b25c-1ac69d7ca848

I've double checked that my Lazer and Nvidia drivers are both up to date. If they're useful, here's the log files from above:

[compressed-logs.zip](https://github.com/ppy/osu/files/13854061/compressed-logs.zip)

Cheers!

**Comment by joshuk** [comment]
Just popping in again to say that this is still an issue. Have experienced it a few times since my last post, just forgot to mention.

https://github.com/ppy/osu/assets/25204957/34c739cc-6a10-434f-8a0a-b0e10762a65d

Latest version of Lazer. Logs:

[compressed-logs.zip](https://github.com/ppy/osu/files/14438409/compressed-logs.zip)

Thanks!

**Reply by peppy** [reply, marked as answer]
This looks like something else on your system doing something every so often, causing GPU load. Not something we can fix from our end. Please trial-and-error until you figure out what it is.

**Reply by joshuk** [reply]
Fair play. I had a suspicion this would be the case, just wanted to double check.

Cheers!
